chain/vm/spec_shim.go

```diff
@@ -56,62 +84,156 @@ func (rs *runtimeShim) shimCall(f func() interface{}) (rval []byte, aerr aerrors
 	}
 }
 
-func (rs *runtimeShim) ValidateImmediateCallerIs(as ...address.Address) {
-	for _, a := range as {
-		if rs.vmctx.Message().From == a {
-			return
-		}
+func (rs *runtimeShim) Message() vmr.Message {
+	var err error
+
+	rawm := *rs.vmctx.Message() // TODO: normalize addresses earlier
+	rawm.From, err = rs.vmctx.LookupID(rawm.From)
+	if err != nil {
+		rs.Abortf(exitcode.ErrPlaceholder, "resolve from address: %v", err)
 	}
-	fmt.Println("Caller: ", rs.vmctx.Message().From, as)
-	panic("we like to panic when people call the wrong methods")
+
+	rawm.To, err = rs.vmctx.LookupID(rawm.To)
+	if err != nil {
+		rs.Abortf(exitcode.ErrPlaceholder, "resolve to address: %v", err)
+	}
+
+	return &rawm
 }
 
-func (rs *runtimeShim) ImmediateCaller() address.Address {
-	return rs.vmctx.Message().From
+func (rs *runtimeShim) ValidateImmediateCallerAcceptAny() {
+	return
 }
 
-func (rs *runtimeShim) Context() context.Context {
-	return rs.vmctx.Context()
+func (rs *runtimeShim) CurrentBalance() abi.TokenAmount {
+	b, err := rs.vmctx.GetBalance(rs.vmctx.Message().To)
+	if err != nil {
+		rs.Abortf(exitcode.ExitCode(err.RetCode()), "get current balance: %v", err)
+	}
+	return b
 }
 
-func (rs *runtimeShim) IpldGet(c cid.Cid, o vmr.CBORUnmarshaler) bool {
-	if err := rs.vmctx.Storage().Get(c, o); err != nil {
-		panic(err) // y o o o o o l l l l o o o o o
+func (rs *runtimeShim) GetActorCodeCID(addr address.Address) (ret cid.Cid, ok bool) {
+	ret, err := rs.vmctx.ActorCodeCID(addr)
+	if err != nil {
+		// todo: notfound
+		rs.Abortf(exitcode.ErrPlaceholder, "%v", err)
 	}
-	return true
+
+	return ret, true
 }
 
-func (rs *runtimeShim) IpldPut(o vmr.CBORMarshaler) cid.Cid {
-	c, err := rs.vmctx.Storage().Put(o)
+func (rs *runtimeShim) GetRandomness(personalization crypto.DomainSeparationTag, randEpoch abi.ChainEpoch, entropy []byte) abi.Randomness {
+	r, err := rs.vmctx.GetRandomness(personalization, randEpoch, entropy)
 	if err != nil {
-		panic(err)
+		rs.Abortf(exitcode.SysErrInternal, "getting randomness: %v", err)
 	}
-	return c
+
+	return r
 }
 
-func (rs *runtimeShim) Abort(code exitcode.ExitCode, msg string, args ...interface{}) {
-	panic(aerrors.Newf(uint8(code), msg, args...))
+func (rs *runtimeShim) Store() vmr.Store {
+	return rs
 }
 
-func (rs *runtimeShim) AbortStateMsg(msg string) {
-	rs.Abort(101, msg)
+func (rs *runtimeShim) NewActorAddress() address.Address {
+	var b bytes.Buffer
+	if err := rs.ImmediateCaller().MarshalCBOR(&b); err != nil { // todo: spec says cbor; why not just bytes?
+		rs.Abortf(exitcode.ErrSerialization, "writing caller address into a buffer: %v", err)
+	}
+
+	var err error
```

Comment:
this declaration can be dropped

Reply:
It's there because `rs.vmctx.StateTree` returns `ActorError`